File: lingting-spring-web/src/main/java/live/lingting/spring/web/scope/WebScope.java
package live.lingting.spring.web.scope;

import java.util.HashMap;
import java.util.Map;

/**
 * @author lingting 2024-03-20 15:02
 */
public class WebScope {

	private final String scheme;

	private final String host;

	private final String origin;

	private final String ip;

	private final String uri;

	private final String traceId;

	private final String requestId;

	private final String language;

	private final String authorization;

	private final String userAgent;

	private final Map<String, Object> attributes = new HashMap<>();

	public WebScope(String scheme, String host, String origin, String ip, String uri, String traceId, String requestId,
			String language, String authorization, String userAgent) {
		this.scheme = scheme;
		this.host = host;
		this.origin = origin;
		this.ip = ip;
		this.uri = uri;
		this.traceId = traceId;
		this.requestId = requestId;
		this.language = language;
		this.authorization = authorization;
		this.userAgent = userAgent;
	}

	public void putAttribute(String key, Object value) {
		this.attributes.put(key, value);
	}

	public <T> T getAttribute(String key) {
		return (T) this.attributes.get(key);
	}

	public String getScheme() {
		return this.scheme;
	}

	public String getHost() {
		return this.host;
	}

	public String getOrigin() {
		return this.origin;
	}

	public String getIp() {
		return this.ip;
	}

	public String getUri() {
		return this.uri;
	}

	public String getTraceId() {
		return this.traceId;
	}

	public String getRequestId() {
		return this.requestId;
	}

	public String getLanguage() {
		return this.language;
	}

	public String getAuthorization() {
		return this.authorization;
	}

	public String getUserAgent() {
		return this.userAgent;
	}

	public Map<String, Object> getAttributes() {
		return this.attributes;
	}

}

File: lingting-spring-web/src/main/kotlin/live/lingting/spring/web/scope/WebScope.kt
package live.lingting.spring.web.scope

import java.util.Locale
import live.lingting.framework.http.header.HttpHeaders
import live.lingting.framework.time.DateTime

/**
 * @author lingting 2024-03-20 15:02
 */
@Suppress("UNCHECKED_CAST")
class WebScope {

    @JvmOverloads
    constructor(
        scheme: String = "",
        host: String = "",
        origin: String = "",
        ip: String = "",
        uri: String = "",
        traceId: String = "",
        requestId: String = "",
        language: String = "",
        authorization: String = "",
        userAgent: String = "",
        headers: HttpHeaders = HttpHeaders.empty(false),
        attributes: MutableMap<String, Any> = HashMap(),
    ) {
        this.scheme = scheme
        this.host = host
        this.origin = origin
        this.ip = ip
        this.uri = uri
        this.traceId = traceId
        this.requestId = requestId
        this.language = language
        this.authorization = authorization
        this.userAgent = userAgent
        this.headers = headers
        this.attributes = attributes
    }

    val time = DateTime.current()
    val scheme: String
    val host: String
    val origin: String
    val ip: String
    val uri: String
    val traceId: String
    val requestId: String
    val language: String
    val authorization: String
    val userAgent: String
    val headers: HttpHeaders
    val attributes: MutableMap<String, Any>

    fun putAttribute(key: String, value: Any) {
        this.attributes.put(key, value)
    }

    fun <T> getAttribute(key: String): T {
        return this.attributes[key] as T
    }

    fun language(): Locale? {
        try {
            if (language.isBlank()) {
                return null
            }
            return Locale.forLanguageTag(language)
        } catch (_: Exception) {
            return null
        }
    }

}
